staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go

```diff
@@ -101,6 +103,75 @@ func (s *Downloader) Download(handler http.Handler, etag string) (returnSpec *sp
 	}
 }
 
+// gvList is a struct for the response of the /openapi/v3 endpoint to unmarshal into
+type gvList struct {
+	Paths []string `json:"Paths"`
+}
+
+// SpecETag is a OpenAPI v3 spec and etag pair for the endpoint of each OpenAPI group/version
+type SpecETag struct {
+	spec *spec3.OpenAPI
+	etag string
+}
+
+// DownloadV3 downloads OpenAPI v3 for all groups of a given handler
+func (s *Downloader) DownloadV3(handler http.Handler) (returnSpec map[string]*SpecETag, err error) {
+	// TODO(jefftree): https://github.com/kubernetes/kubernetes/pull/105945#issuecomment-966455034
+	// Move to proxy request in the aggregator and let the APIServices serve the OpenAPI directly
+	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
+	handler = http.TimeoutHandler(handler, specDownloadTimeout, "request timed out")
+
+	req, err := http.NewRequest("GET", "/openapi/v3", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Accept", "application/json")
+
+	writer := newInMemoryResponseWriter()
+	handler.ServeHTTP(writer, req)
+
+	switch writer.respCode {
+	case http.StatusNotFound:
+		// Gracefully skip 404, assuming the server won't provide any spec
+		return nil, nil
+	case http.StatusOK:
+		groups := gvList{}
+		aggregated := make(map[string]*SpecETag)
+
+		if err := json.Unmarshal(writer.data, &groups); err != nil {
+			return nil, err
+		}
+		for _, path := range groups.Paths {
+			reqPath := fmt.Sprintf("/openapi/v3/%s", path)
+			req, err := http.NewRequest("GET", reqPath, nil)
+			if err != nil {
+				return nil, err
+			}
+			req.Header.Add("Accept", "application/json")
+			openAPIWriter := newInMemoryResponseWriter()
+			handler.ServeHTTP(openAPIWriter, req)
+			var spec spec3.OpenAPI
+			// TODO|jefftree: For OpenAPI v3 Beta, if the v3 spec is empty then
+			// we should request the v2 endpoint and convert it to v3
+			if len(openAPIWriter.data) > 0 {
+				err = json.Unmarshal(openAPIWriter.data, &spec)
+				if err != nil {
+					return nil, err
+				}
+				etag := writer.Header().Get("Etag")
```

Comment:
What does this return if there's no etag? I can see above that you're comparing etags, but I suspect this returns an empty string if it doesn't exist, and empty strings will compare as equal.

Reply:
On the other hand, since v3 will ALWAYS have etags enabled (compared to v2 where the feature was added later on), you can maybe assume that it's there?

Reply:
The etag should be guaranteed to exist if the APIService uses the OpenAPI v3 handlers that we provide. Is that a safe assumption to make? If the APIService is old and does not publish v3, then we don't care about the etag because `len(openAPIWriter.data)` will be 0.